packages/doenetml-worker-rust/lib-doenetml-core/src/graph/directed_graph.rs

```diff
@@ -168,28 +175,96 @@ use iterators::*;
-    /// may yield the same node multiple times.
-    pub struct DescendantIterator<'a, Node> {
+    /// Iterate through all edges of a node an its descendants.
+    /// This returns edges represented with `DirectedGraph`'s internal representation.
+    ///
+    /// **For internal use**. Only use this function if you know what you're doing.
+    pub struct DescendantEdgeIteratorRaw<'a> {
+        edges: &'a [Vec<usize>],
+        remaining_indices: Vec<(usize, usize)>,
+        descendant_iterator_raw: DescendantIteratorRaw<'a>,
+    }
+    impl<'a> DescendantEdgeIteratorRaw<'a> {
+        pub fn new(num_nodes: usize, edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            let descendant_iterator_raw =
+                DescendantIteratorRaw::new(num_nodes, edges, start_index);
+            let remaining_indices = edges[start_index]
+                .iter()
+                .map(|&to| (start_index, to))
+                .collect();
+            Self {
+                edges,
+                remaining_indices,
+                descendant_iterator_raw,
+            }
+        }
+    }
+    impl<'a> Iterator for DescendantEdgeIteratorRaw<'a> {
+        type Item = (usize, usize);
+        fn next(&mut self) -> Option<Self::Item> {
+            if let Some(edge) = self.remaining_indices.pop() {
+                Some(edge)
+            } else {
+                // We've exhausted our current supply of edges. Move to the next descendant.
+                let next_descendant = self.descendant_iterator_raw.next()?;
+                self.remaining_indices.extend(
+                    self.edges[next_descendant]
+                        .iter()
+                        .map(|&to| (next_descendant, to)),
+                );
+                self.next()
+            }
+        }
+    }
+
+    pub struct DescendantEdgeIterator<'a, Node> {
         nodes: &'a [Node],
+        descendant_edge_iterator_raw: DescendantEdgeIteratorRaw<'a>,
+    }
+    impl<'a, Node> DescendantEdgeIterator<'a, Node> {
+        pub fn new(nodes: &'a [Node], edges: &'a [Vec<usize>], start_index: usize) -> Self {
+            Self {
+                nodes,
+                descendant_edge_iterator_raw: DescendantEdgeIteratorRaw::new(
+                    nodes.len(),
+                    edges,
+                    start_index,
+                ),
+            }
+        }
+    }
+    impl<'a, Node> Iterator for DescendantEdgeIterator<'a, Node> {
+        type Item = (&'a Node, &'a Node);
+        fn next(&mut self) -> Option<Self::Item> {
+            let (from, to) = self.descendant_edge_iterator_raw.next()?;
+            Some((&self.nodes[from], &self.nodes[to]))
+        }
+    }
+
+    /// Iterate through the descendants of a node but return
+    /// the `DirectedGraph`'s internal representation of each node.
+    ///
+    /// **For internal use**. Only use this function if you know what you're doing.
+    pub struct DescendantIteratorRaw<'a> {
```

Comment:
Mark this as private?

Reply:
It has to be public since it is used by a parent..., but I can certainly make it `pub(crate)`

Reply:
The current code still complies for me if I get rid of `pub` on the two `Raw` structs.